### ISSUE TITLE
Flatpak support

### ISSUE THREAD

**Issue by vchernin** [waydroid/waydroid#64]
https://www.flatpak.org/

This would make it hopefully easier to install on non Debian-based distros.

Since Flatpak is a container based system, I'm not sure how easy this is given Waydroid is itself a container.

**Comment by mmcnutt** [comment]
Waydroid relys on binder and ashmem, which most distros dont enable in their kernels by default. There's probably no reason that you could package libgbinder and the user space components, but for now the stumbling block for people looking for easy Waydroid would be the kernel.

**Comment by kevinsmia1939** [comment]
Please make app suggestion in flathub forum.

**Comment by allaeddineomc** [comment]
> Waydroid relys on binder and ashmem, which most distros dont enable in their kernels by default. There's probably no reason that you couldn't package libgbinder and the user space components, but for now the stumbling block for people looking for easy Waydroid would be the kernel.

maybe making a waydroid flatpak would make more people notice it and want to use it since they will be confident that installing it is easy and won't break their system or fill it with garbage , than it will be a matter of time for distros to start adding binder and ashmem to their kernels because a distro that can't run mainstream software is a bad distro and distro maintainers know this very well

**Comment by vchernin** [comment]
Unfortunately doing this will likely be a non-trivial amount of work, because Flatpak itself is a sandboxed container. It's possible Waydroid would have to target a Flatpak container instead of the LXC container for it to work. Otherwise a container within a container is not really feasible.

Still I would be happy to help try write a Flatpak if there's confirmation it's theoretically possible.

**Comment by allaeddineomc** [comment]
> Unfortunately doing this will likely be a non-trivial amount of work, because Flatpak itself is a sandboxed container. It's possible Waydroid would have to target a Flatpak container instead of the LXC container for it to work. Otherwise a container within a container is not really feasible.
> 
> Still I would be happy to help try write a Flatpak if there's confirmation it's theoretically possible.

no that's not the case , waydroid can still use lxc as it is when put into a flatpak running a container inside a sandboxed envirenment isn't anything new , take for example anbox snap , and also note that running a container inside another container have no performance hit what soever since containers are just an easy to mange chroots that run natively

**Comment by kevinsmia1939** [comment]
@allaeddineomc 
Flatpaking anbox was discuss here.
https://github.com/anbox/anbox/issues/62

It is not trivia to do it and no one manage to do it.

**Comment by allaeddineomc** [comment]
> @allaeddineomc
> Flatpaking anbox was discuss here.
> [anbox/anbox#62](https://github.com/anbox/anbox/issues/62)
> 
> It is not trivial to do it and no one manage to do it.

guys who do snap won't do flatpak just because they hate it , same for flatpak guys won't make a snap because it's not really universal and very slow , they haven't done it not because it's hard but because they don't want

**Comment by nsrosenqvist** [comment]
Just came across this bit of news that might be of interest, highlighting Flatpak's improved sub-sandbox support: https://www.phoronix.com/scan.php?page=news_item&px=Flatpak-1.12-Released

**Comment by krillin666** [comment]
Any update on this ? :grin:

**Comment by rugk** [comment]
Indeed it would be nice if you could publish this as a [flatpak](https://flatpak.org/) on [flathub](https://flathub.org/) e.g.
Flatpaks are a new software distribution mechanism for Linux distros, can thus installed on any distro and are easy to update. 
Here is [how to get started](http://docs.flatpak.org/en/latest/getting-started.html).

BTW, as teh snap discussion sparked, I personally would prefer flatpak. In contrast to snap, you can also self-host it, so you stay in control, and it is widely supported. (snap is not so nice to setup in many distros and snap's security depends on AppArmor, which is not always available in many distros)
Also – in contrast to snaps – flatpaks do not only claim to be distro-independent, but [actually are](https://kamikazow.wordpress.com/2018/06/08/adoption-of-flatpak-vs-snap-2018-edition/).
Furthermore even [Linux Mint criticizes snap](https://blog.linuxmint.com/?p=3906), because you cannot self-host a snap server as it is proprietary and (thus) also cannot modify the packages that are served centralized by Ubuntu.

**Comment by Fuseteam** [comment]
But snap __is__ [distro independent people install it on distros that don't ship it](https://snapcraft.io/core). 
Linux mint's blog post is full of misinformation about snap

And noone in this entire thread has given any real technical details what it would require to setup and lxc container as root and a session as the user as a flatpak.
Snaps and flatpaks simply are not comparable both have their own goals and purposes. Flatpak is simply not fit for waydroid

Anbox is years old, [the author is not against commit that enable flatpak support](https://github.com/anbox/anbox/issues/62#issuecomment-294151113), and it has been noted by [prominent flatpak contributor](https://github.com/anbox/anbox/issues/62#issuecomment-356383026) that flatpaks **cannot** get root access, something both anbox and waydroid need to set up their container manager

**Comment by TheEvilSkeleton** [comment]
> But snap **is** [distro independent people install it on distros that don't ship it](https://snapcraft.io/core).

Err not really. Firstly, Snap hard depends on systemd so any distro that does not use systemd, e.g. Gentoo, Artix, Devuan, etc., are very likely to not work with Snap.

Secondly, Snap relies on AppArmor to sandbox, so if you use Fedora Linux or any dsitro that does not use AppArmor, then the sandboxing is practically nonexistent, so it's really not distro independent.

But I digress, Waydroid Flatpak is unlikely to happen because of the reasons you have mentioned. Flatpak is too restrictive so for now it's impossible, unless we find workarounds.

**Comment by Fuseteam** [comment]
fwiw snap works fine on fedora and centos both from the red hat line, as seen from the stats people are using it. apparmor and selinux aren't particularly mutually exclusive both can work together.
there are ways to get working on gentoo (it's even mentioned in the link you quoted xd) systemd isn't exactly a monolithic thing in fact ubuntu 16.04, that ran snaps, still shipped upstart :p ok a mix of systemd and upstart point is not the __entirety__ of systemd is need, just a specific component. if i had to guess which, its systemd mounting mechanism.........but i'm straying offtopic
i honestly don't think we should force waydroid into flatpak, flatpaks are best suited for actual desktop software. and they work great for that, i see no reason for flatpaks to try and match snap in functions. the two are made for different usecases, and that's fine

waydroid is more suited as a traditional package anyway, where it can get the access it needs to run. if we have to go to a universal package there is only one suitable format. but it appears the community won't appreciate that effort

**Comment by TheEvilSkeleton** [comment]
> fwiw snap works fine on fedora and centos both from the red hat line, as seen from the stats people are using it.

How do you define "fine"? The sandboxing is literally nonexistent because Snap relies on AppArmor, which Fedora doesn't use. And that's anything but fine. This isn't Android where only one MAC is used. We have AppArmor, SELinux and more. Furthermore, distributions that don't even have a MAC pre-installed, like Arch Linux, aren't benefiting from the sandboxing either, unless they install AppArmor.

> apparmor and selinux aren't particularly mutually exclusive both can work together.

I don't know of a mainstream distro that uses both, so whether what you're saying is true or false, is impractical at best.

> there are ways to get working on gentoo (it's even mentioned in the link you quoted xd) systemd isn't exactly a monolithic thing

Gentoo lets you choose other init systems/service managers, including systemd. However, by default, it uses OpenRC. Many Gentoo users use Gentoo simply because systemd is an option and not an obligation. But my point still stands: you cannot run Snap on systems that aren't using systemd.

> in fact ubuntu 16.04, that ran snaps, still shipped upstart :p

[Okay?](https://snapcraft.io/docs/services-and-daemons)

Really, there are very little reasons for using Snap on the desktop. It's a server package manager poorly adapted to the desktop.

**Comment by ShadowJonathan** [comment]
> Really, there are very little reasons for using Snap on the desktop. It's a server package manager poorly adapted to the desktop.

Interestingly, i disagree, as my experience with snaps has been abysmal for server application management, simply because it does not allow update control. The developers being more and more hostile and dismissive the more i press on this only seems to prove they think they know best for their users, but for server software, administrators simply need full control, which snap doesn't get.

I thought the saving grace would be that snaps "at least" would be more useful on the desktop, for user applications, and left it at that, but alas.

**Comment by Fuseteam** [comment]
I disagree my experience with snap has been fine, "update control" is less of an issues one you learn of snap's channel system. Also the app awareness system that has been testing.

I won't comment on whether or not sandboxing exists on distros without apparmor, all i know is that it is used on those system and people continue to use it on those systems. My main point about gentoo and possibly other systemd-less systems is that snap doesn't depend on systemd as a whole. It depends on specific pieces of systemd which can be set up without pulling the rest, people are actually doing that out there. Additionally the codebase for snapd, the core of the package management system, is opensource. If people really wanted to they could make it work without apparmor and without systemd.

But let's not turn this into a snap bashing thread, the thread is about flatpak support. And that simply isn't feasible.

**Comment by TheEvilSkeleton** [comment]
> My main point about gentoo and possibly other systemd-less systems is that snap doesn't depend on systemd as a whole(systemd itself isn't a monolithic thing). Snap simply depends on specific pieces of systemd, probably the part that mounts the squashfs images as loop mount devices at boot time, which can be set up without pulling the rest, people are actually doing that out there. Additionally the codebase for snapd, the core of the package management system, is opensource. If people really wanted to they could make it work without apparmor and without systemd.

So what are you getting at and why does it matter to go in the specifics? It's like saying "Flatpak doesn't rely on the Linux kernel as a whole, only some parts". Sure, it's technically correct but at the end of the day, you can't use Flatpak outside of Linux, just like you cannot use Snap outside of systemd.

My only point was that it's not as universal as people may think because it hard depends on systemd for usability and AppArmor for the sandboxing. I'm unsure why you are defending Snap, when the issues are pretty clear as to why it's not a good option/alternative.

**Comment by ShadowJonathan** [comment]
I think this issue comes down to two things;
1. Support for flatpak packages to require a module to be installed in/for the kernel.
2. Ability to distribute dkms(?) next to flatpak packages.

I think both could be put forward better with an RFC in https://github.com/flatpak/flatpak, what do y'all think?

**Comment by TheEvilSkeleton** [comment]
Looks good. If you don't get any response, you can chat with them at https://matrix.to/#/#flatpak:matrix.org.

**Comment by gdonval** [comment]
Just one thing because people seem to mix things up easily...

> It is not trivial to do it and no one manage to do it.

It is. It's just that no one cleary stated the obvious and acted upon it: waydroid and anbox are leaky as hell. They are dependent on special kernels and special kernel configuration and special boot environment.

Once that fact is accepted, it turns out that shipping a kernel with a program is surprisingly easy: you provide a VM. Gnome Boxes for instance ships with qemu, it works well and is self-contained so this much is trivial. With qemu-kvm, there should me almost no CPU performance loss and with virtio-gpu, gpu acceleration should be decent.

Of course, it does not mean waydroid as a project has to be shipped like that.

**Comment by ShadowJonathan** [comment]
> you provide a VM

I don't think that is specifically neccecary, I was reminded of [User Mode Linux](https://en.wikipedia.org/wiki/User-mode_Linux), where a user can run a linux kernel in usermode as a seperate program. (Now its in the [main source tree](https://www.kernel.org/doc/html/v5.9/virt/uml/user_mode_linux.html))

**Comment by Fuseteam** [comment]
> It is. It's just that no one cleary stated the obvious and acted upon it: waydroid and anbox are leaky as hell. They are dependent on special kernels and special kernel configuration and special boot environment.

the "special" kernel your referring to are likely the ashmem and binder kernel modules, these are requirements of android not of anbox nor waydroid. if you believe "it's trivial to do", by all means do it. this is an 5 year old open issue, discussions have been had in that time. the technically inclined have concluded that it's not technically feasible with to create a flatpak as it stands. no matter how many people say "flatpak +1", "flatpak is awesome", "flatpak is the future", "it's trivial to do, why isn't this done yet" if nobody stands up and actually do something about it it will not be done.

p.s. afaict if we're going into vm territory, we're better off putting lineage os in a kvm and be done with it

**Comment by gungwald** [comment]
We should not need to have a separate package tool for every different kind of app. I just want to install apps. I shouldn't have to deal with 5 different package tools to get the apps I want installed. What's next, are we going to build a separate package tool for every app? This situation is ridiculous. There should be only one package tool. That is what developers should be demanding, one package tool that can install any kind of app. The snap server is proprietary so that eliminates snaps as an option. Flatpak is the standard. If there are problems that prevent an app from being packaged as a flatpak, those issues should be communicated to the flatpak developers so they can provide the necessary features.

**Comment by Fuseteam** [comment]
what the heck are you talking about @gungwald flatpak simply is a technology made to install something like waydroid. by your own logic, waydroid itself is a package manager. 
`waydroid app install <name-of-app>.apk` is [a thing](https://docs.waydro.id/usage/install-and-run-android-applications) just like `apt install <name-of-app>.deb` is a thing, give or take paths

flatpak is a technology made to deliver desktop apps, and honestly it is good at what it does because it does one thing and does it well: deliver deliver apps, and it should stay that way :P snap tries to do much more than that, and you see the sentiment as a result of that :P

Stop trying to shove flatpaks and snaps in the same box, they don't fit. they are different technologies that do different things well

**Comment by kon14** [comment]
@Fuseteam Why are you even so irritated over this?
If Waydroid maintainers feel like this issue is not worth keeping open they can always just close it themselves.

> Stop trying to shove flatpaks and snaps in the same box, they don't fit. they are different technologies that have their own purposes and uses

Somehow I'm willing to bet none of the people subscribed to this thread care enough to place Snaps in any sort of box.
Expect maybe for that very same garbage bin currently occupied by Mir and every single other case of abandonware Canonical NIH projects.

**Comment by Fuseteam** [comment]
I'm irritated because people keep asking something of a technology, that it just wasn't designed for.
Flatpak does things well, but isn't a snap alternative. Just like snap isn't a flatpak alternative. They are different technologies for different purposes.

Speaking of NIH, snap was already a thing, then redhat started flatpak, upstart was already a thing, then redhat started systemd ;)
Nothing wrong with that, sometimes tools don't suit particular needs at a particular time. Time being of the essense it is often easier and quick to build an alternative ;)
P.s. Mir is a wayland display server very similar in architecture to xorg, in practice mutter, kwin and friend could stay window managers on top a wayland compositor, instead of reimplementing a wayland compositor of their own ;)

Long story short; [flatpak cannot be used for waydroid](https://github.com/anbox/anbox/issues/62#issuecomment-356383026). Both anbox and waydroid require to run a part that needs root. Flatpak does not and should not provide such access, flatpak having a clear separation between a application that runs with root privileges and an applications that has zero root privileges is a **good** thing, stop asking to kill a good feature

**Comment by soredake** [comment]
https://github.com/flatpak/flatpak/issues/2787

**Comment by kon14** [comment]
> Speaking of NIH, snap was already a thing, then redhat started flatpak, upstart was already a thing, then redhat started systemd ;)

@Fuseteam Snap was started roughly at the same period as Flatpak and wasn't even originally marketed as a desktop or universal package format, just like OpenRC was started a bit after upstart.
And yet Snap never really made it big on the desktop. Insisting it's magically going to work out one day is still a huge case of NIH regardless of whether their project was started a dozen weeks or a dozen decades prior to anything else.

You can argue over their purpose all you want, but ultimately, if packaging Waydroid as a Flatpak is something Flatpak and Waydroid devs (or even the community) are both interested in, it's eventually going to happen.

I'd suggest we stop spamming this thread with any such pointless talk or "+1 for Flatpak" demands any further. I know will.

**Comment by Fuseteam** [comment]
Nobody saying snap "will magically work one day" snap has it's purpose but elaborating on that is off topic for this thread. My point is snap isn't an alternative to flatpak and flatpak isn't an alternative to snap. They aren't "competing" formats. They are different formats for different purposes. You can't replace one with the other.
Both have their place, and i use both according to my needs. 

Flatpak simply isn't a technology that can be used to package waydroid. This is elaborated in the post i linked, forcing flatpak to able to package waydroid would mean killing one of the features of flatpak that makes suitable for purposes that snap isn't suitable for.

The unix philosophy that so many cite against systemd "do one thing, and do it well" is flatpak has going for it, it does one thing, deliver unprivileged desktop applications, and it does it well. Snap tries to do too many things which resulted in being overengineerd for the desktop usecase. So in short I'm saying flatpak isn't snap, keep it that way.

**Comment by rugk** [comment]
IMHO Flatpaks job is to package (and sandbox and make it universally executable etc.) desktop applications. And I consider Waydroid a desktop application, as it clearly has a visible UI. That it technically may not work yet/is hard, is a totally different topic, but I stand by the idea that it would be a good idea/fit into flatpak if it could be installed like that…

**Comment by Fuseteam** [comment]
Waydroid is far from a "desktop application". Waydroid is a LXC container manager to run Android on top of Linux in order to install and run Android application on desktop Linux.
in a way waydroid is a package manager on its own, `waydroid app install/launch/list/remove (name.apk)` is [a thing already](https://docs.waydro.id/usage/install-and-run-android-applications), all it needs now is a repo configured for it to pull apk packages from and boom android package manager xd

Waydroid is as much a "desktop" Application as Docker is a "desktop" application(which also cannot be packaged as a flatpak for the same reason).

**Comment by electrikjesus** [comment]
This should help fill some of those holes https://github.com/waydroid/waydroid-package-manager

**Comment by rugk** [comment]
Okaaay though talking about things like VMs GNOME boxes can run in a flatpak. I did not really got *how* they managed to do thus, but in any case it works.🙃🙃

**Comment by Fuseteam** [comment]
Then unlike waydroid, docker and kvm, gnome boxes doesn't need root access. :p

**Comment by 12people** [comment]
@Fuseteam It's true that, conceptually, Waydroid is not a desktop application. Conceptually, it's more akin to WINE.

There are Flatpaks of desktop applications that provide a GUI for WINE — e.g. https://flathub.org/apps/details/com.usebottles.bottles .

While I understand there are currently limitations to Flatpaks that don't allow Waydroid to be packaged up the same way, I'm sure the Waydroid and Flatpak teams can work together to overcome those limitations.

**Comment by Fuseteam** [comment]
@12people while conceptually waydroid is closer to wine than a desktop application, architechtually it isn't even in the same ballpark. 
wine can be seen an library that can be used to run windows applications. bottles is desktop application that utilizes wine.
waydroid is way closer to docker containers than to wine. in fact it __is__ a container that enables an android application. trying to get waydroid to be packaged in a flatpak is like trying to transport a truckload of sand with a ferrari. 
just like trying dumping a truckload of sand on a ferrari destroys the ferrari, forcing flatpak to package waydroid will kill off the good features of flatpak.

stop trying force everything in flatpaks, it won't do anyone any favors. flatpaks have usecases they are good for, use flatpaks for those use cases instead.

flatpaks enable users to install applications on immutable filesystem, that's a good thing.
flatpak enable users to application without resorting to root, that's a good thing
waydroid requires root and not root permissions to function, flatpak cannot do that without breaking flatpak's security mechanisms that would be a bad thing

stop asking to break the ferrari so you can use it as daf truck

**Comment by carlocastoldi** [comment]
What @Fuseteam says seems completely reasonable to me, tbf.

That said his words made me wonder, since Waydroid is a container (not too dissimilar from Snap & Flatpak technologies, afaik), if it would be possible to interface Waydroid's app permissions requests—and the Android system in general—with [XDG Desktop Portals](https://github.com/flatpak/xdg-desktop-portal) when asking for host's resources.
There are already (at least) two sandbox environments that make use of this new & improving technology. Why not adding a third?

**Comment by Fuseteam** [comment]
@carlocastoldi now that's nice point i too would love to see that, waydroid as a "format" for installing android applications sound very enticing to me. the more we can integrate waydroid with linux tools the better. maybe this can be first step to pull the android app ecosystem into glinux.

from my limited understanding we would need to write a xdg desktop portal backend for it that would then interface with the android lxc container 🤔

**Comment by carlocastoldi** [comment]
@Fuseteam that is pretty much the same idea that a maintainer of XDG Portals suggested here: https://github.com/flatpak/xdg-desktop-portal/issues/637#issuecomment-956498729

But he then proceeded suggesting to use a Flatpak per Waydroid's application, tho

**Comment by Fuseteam** [comment]
@carlocastoldi that's an interesting approach, to me it sounds like packaging android apps as flatpaks so a new approach to running android apps separate but perhaps not that separate from waydroid. 

His comparison with valve's pressure vessels showcase how waydroid differs in a fundamental way. 
afaict Steam games don't require root permissions to run (much like wine), however waydroid, and by extension anbox (or is that the otherway around?), requires root permissions to start the container while the session **has** to be started with the regular user.

if i understand correctly there _has_ to be a clear boundary between the container and the session or it won't work as expected

**Comment by reesericci** [comment]
Just spitballing - is there any way to get Waydroid's containers to be built on top of Podman? Podman doesn't require root access, and such it could be run in Flatpak. I only have surface-level knowledge of these technologies, so this may be a stupid question.

**Comment by Fuseteam** [comment]
@reesericci that's a valid question, sadly it is not that lxc containers requires root access, it's that running the android system in lxc containers requires root access.

however you mentioning podman does makes me think; what if we take a step back and look at our goal? our goal and likely the goal of everyone in this issue is to install a android container as a flatpak. in that case there is one we could try: https://github.com/remote-android/redroid-doc
this would be able to work with podman. how well it works and how well we can integrate it is another question. tho as they say nothing ventured, nothing gained

**Comment by schklom** [comment]
> that's a valid question, sadly it is not that lxc containers requires root access, it's that running the android system in lxc containers requires root access.

Does it actually require reading or modifying root files?
I run quite a few Docker containers that "require" root using [Rootless Docker](https://docs.docker.com/engine/security/rootless/) and they work perfectly fine. Rootless Docker uses namespaces to map host user IDs to multiple container user IDs, root on container is translated to the user running Docker. Access to (host) root files is prevented by the host.
No idea about Podman though.

**Comment by Fuseteam** [comment]
@schklom read what i said again
> sadly it is **not** that lxc requires root........**running android inside of lxc requires root**

also consider reading the context, just a few posts up i explained that
> waydroid..... requires root permissions to start the container while the session **has** to be started with the regular user.

i guess you could say running waydroid requires **2 users**: the root user and the regular user
the container is started by the root user while the session is started by the regular user

**Comment by orowith2os** [comment]
Would it be possible to make Waydroid use podman for containers, rather than LXC? It would still require that the kernel has the required modules, I think, but would allow for running Waydroid without root (and a Flatpak can just use the host podman configuration for managing containers). 

There might also be the possibility of using bwrap, as the Steam flatpak uses it for the Steam Linux Runtime and Flatpak spawns a separate sandbox for it, so that there's no sandbox-ception.

That said: I'm not familiar with how, exactly, Waydroid works, and have only used podman for using multiple Linux distros at a time. I only have a basic idea of how sandboxing works. Take my ideas with a grain of salt.

**Comment by mmcnutt** [comment]
@orowith2os I have attempted to get waydroid running in toolbox (which is essentially a frontend for podman) on Fedora Kinoite, and the stumbling block is always exposing the binderfs to the container. I understand this isnt exactly what you are asking as its still an LXC container inside the podman container, but I imagine even dropping android into a podman container is still going to have the same binderfs limitation.

My ultimate goal in trying was avoiding having to layer the waydroid packages on my ostree. Trying to get wd into a flatpak would achieve the same thing too

**Comment by Fuseteam** [comment]
for what it is worth: waydroid requiring root is not due to lxc. it is due to the way android works.

Android needs the system to run as root and the session to run as the actual user. Remember We're running an operating system in a container. Android is a multiuser operating system. As far as i've understood android working properly depends on there being a root user and a normal user, fully separated from each other in order to work.

This is why if you do `sudo waydroid session start` waydroid will break.

**Comment by orowith2os** [comment]
I can have a (seemingly) rootful environment, inside of a rootless container. I don't think Android needing to run as root and the session as the user will change anything. All I can see that needs changing is for some bwrap/podman-specific things, and then it could work properly, although I'm not all that familiar with how it would be done.

**Comment by Fuseteam** [comment]
there are a lot of ways to have a fake "rootful" environment, that doesn't necessarily mean android will behave in it. i suppose this could be explored through the redroid project

**Comment by pktiuk** [comment]
> waydroid's goal, which is to run as many android apps as seamlessly as possible on desktop linux

Easy to setup and use package for my distro is in my opinion a pretty important part of a seamless experience. (nor now only debian-based distros are supported)

It is a very good idea to get to know from redroid  what is needed for android working properly in container which may be helpful with packaging it info flatpak.

**Comment by Fuseteam** [comment]
> Easy to setup and use package for my distro is in my opinion a pretty important part of a seamless experience. (for now only debian-based distros have official packages)

sure but that also means that most apps just work, forcing android in a permission scheme it was not build for will break apps. Why don't you try [redroid](https://github.com/remote-android/redroid-doc)? perhaps that could be wrapped in a waydroid like wrapper?

also how much easier can you get than
```
sudo deb-get install waydroid
sudo systemctl start waydroid-container-manager
systemctl start --user waydroid-session-manager
```
huh? that last one isn't a thing yet? guess i know what my first pr will be........

**Comment by pktiuk** [comment]
> sure but that also means that most apps just work, forcing android in a permission scheme it was not build for will break apps.

I know that flatpak package will be not very reliable at the very beginning, but in the long time it may be worth the effort to have it done.   
Do we have for now any single thing (interface/API/permission) which we can point at and tell that it is blocker and packaging into flatpak is not possible until they (people behind flatpak or bwrap) will fix it?  
It would be a good justification to leave flatpak packaging for now.   

> Why don't you try [redroid](https://github.com/remote-android/redroid-doc) first?

Redroid is a similar case. Docker is not the way software should be provided to end-users.

**Comment by orowith2os** [comment]
If permissions are an issue, the Flatpak can be unlocked a lot more than you'd think, since you can just enable the permissions in the manifest. I think it would be ideal to first figure out how to get Android working inside of bwrap, and THEN work on making a Flatpak. AFAIK the current method Waydroid uses is similar to bwrap, so should be easy enough.

**Comment by Fuseteam** [comment]
> I know that flatpak package will be not very reliable at the very beginning, but in the long time it may be worth the effort to have it done. Do we have for now any single thing (interface/API/permission) which we can point at and tell that it is blocker and packaging into flatpak is not possible until they (people behind flatpak or bwrap) will fix it? It would be a good justification to leave flatpak packaging for now.

the issue is, as stated multiple times in the thread, in order for android to work correctly you need a root user and a session user. 
flatpaks cannot run one process as root and one process as a regular user, as it is made for desktop applications. Desktop applications shouldn't be doing this kind of sourcery, and shouldn't need to.

> Redroid is a similar case. Docker is not the way software should be provided to end-users.

people are suggesting packing android in a podman container, this essentially what redroid already does.

Since nobody seems to have given redroid a try yet, let give you guys some motivation to give it a try:

i'll go on a limp and say that redroid has no hardware access: no camera, no gps, no bluetooth, no mic, no sensors whatsoever (something waydrotd __does__ have).

why do i expect that? because it is my understanding that on linux, access to the hardware mentioned is managed by processes started by the root user.

now......anyone care to confirm or deny this?

**Comment by orowith2os** [comment]
Again, a Flatpak can be unlocked to allow host access (camera, bluetooth, mic, host filesystem), and it can be in a seemingly root environment, so that should work from my understanding. The only way to know if it would actually WORK is if someone were to make a bwrap-compatible image of Android and give it a try. Judging from what I've been told, it only needs to disable a few lines of code to start.

**Comment by Fuseteam** [comment]
if you believe you can make a packaged version using the redroid container work, then by all means do try ;)

**Comment by lifeiscontent** [comment]
For those of you wanting to run Andriod games on Steam Deck via Waydroid, it's currently possible, but a pretty hairy process. https://gitlab.com/popsulfr/steam-deck-tricks/-/tree/main#android-via-waydroid

**Comment by darkmastermindz** [comment]
> For those of you wanting to run Andriod games on Steam Deck via Waydroid, it's currently possible, but a pretty hairy process. https://gitlab.com/popsulfr/steam-deck-tricks/-/tree/main#android-via-waydroid

This is the link I've been looking for! Thanks. Although flatpack support would be amazing to bring Waydroid to target the steam deck in a simpler manner.

**Comment by aleasto** [comment]
Running android in Flatpak would be an entirely different project.
Flatpak is not just a packaging spec. Waydroid works by sharing the host device nodes to the android container, while flatpak has the exact opposite intention.
Not going to happen here.

**Comment by orowith2os** [comment]
*psssst, you can close something as not planned instead of completed*

**Comment by Thesola10** [comment]
Fwiw Waydroid runs fine under Bubblewrap on my Deck, so making a flatpak with "all-devices" access _should theoretically work_

**Comment by RaptaG** [comment]
**Disclaimer:** I'm really sorry for anybody subscribed to this issue getting notified.

Is there anybody which would be willing to create and maintain a sandboxed, flatpak/flathub version of Waydroid? This is not a direct question to the maintainers/developers of Waydroid but to anyone with the motivation, skills and time to maintain such a project.

**Comment by westurner** [comment]
Why would it need to bypass the Flatpak sandbox security features?

**Comment by Wladefant** [comment]
Here is a request on flathub: https://discourse.flathub.org/t/waydroid-android-apps-on-linux-request/2040

**Comment by Fuseteam** [comment]
guys, please read the thread before posting. It has been discussed, including why it is unfeasible to package waydroid as a flatpak. That's the least amount of research you can do, there's a reason this issue is closed.

Waydroid works as well as it does because how it is build.
flatpaks work as well as they do because they strongly defined their scope.

The way waydroid is build does not fit in the scope flatpak defined.

changing the scope of flatpak will ruin what it is good at, so that shouldn't be done.

changing the way waydroid is build will result in a whole new project and will likely function differently (see dock-droid which _does_ fit in the scope of flatpak)

**Comment by Fuseteam** [comment]
@westurner that is a question for [dock-droid](https://github.com/sickcodes/dock-droid) and out of scope for this thread

**Comment by rugk** [comment]
> see dock-droid which does fit in the scope of flatpak

Ah, okay, then why does nobody ask there for a flatpak?

Fixed that so everyone who has subscribed here can unsubscribe and subscribe there ( :wink: ):
https://github.com/sickcodes/dock-droid/issues/22

**Comment by Twig6943** [comment]
Is it possible to package waydroid as an AppImage instead using [conty](https://github.com/Kron4ek/Conty)

**Comment by Twig6943** [comment]
> Is it possible to package waydroid as an AppImage instead using [conty](https://github.com/Kron4ek/Conty) ???

@ivan-hc might be able to pull this off (knows pretty much everything there is to know about appimages)

**Comment by pktiuk** [comment]
I think another issue (or thread in discussions) would be a good starting point for discussion about AppImages.

This one is about flatpak and it should stay this way

**Comment by Fuseteam** [comment]
as mentioned in this thread, waydroid is a privileged application. so it cannot work with any other format than snap, can we please not necro that discussion again

**Comment by Twig6943** [comment]
What about a snap package then?

**Comment by Fuseteam** [comment]
Someone would need to pr the snapcraft.yml, feel free to raise an issue about it/work on it

**Comment by ghost** [comment]
Hey, those android kernel modules got a rust rewrite and will be in Linux kernel in 6.18. Do you think this opens the door for a potential flatpak?
https://rust-for-linux.com/android-binder-driver

**Comment by Fuseteam** [comment]
no, that still does not address the limitations outlined further up in this issue

**Comment by Twig6943** [comment]
Kinda bumping this as this was never posted to this issue; https://github.com/pkgforge-dev/Waydroid-AppImage

This still requires the kernel module to be installed on the host. And is waiting on this issue to be solved before being upstreamed.
https://github.com/waydroid/waydroid/issues/2266

**Comment by bojogon7** [comment]
Maybe consider https://cpak.it/ it uses native Linux kernel sandboxing (no bubblewrap and apparmor) and doesn't require systemd

Edit: 
Made separate issue #2404

**Comment by Fuseteam** [comment]
@bojogon7 cpak is not flatpak— please don't necro unrelated issues, create a new one
